vm/actor/src/builtin/miner/deadline_state.rs

```diff
@@ -0,0 +1,998 @@
+        let no_early_terminations = self.early_terminations.is_empty();
+        Ok((result, !no_early_terminations))
+    }
+
+    pub fn pop_expired_partitions<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        until: ChainEpoch,
+        quant: QuantSpec,
+    ) -> Result<(BitField, bool), Box<dyn StdError>> {
+        let mut expirations = BitFieldQueue::new(store, &self.expirations_epochs, quant)?;
+        let (popped, modified) = expirations
+            .pop_until(until)
+            .map_err(|e| ActorError::downcast_wrap(e, "failed to pop expiring partitions"))?;
+
+        if modified {
+            self.expirations_epochs = expirations.amt.flush()?;
+        }
+
+        Ok((popped, modified))
+    }
+
+    pub fn terminate_sectors<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        sectors: &Sectors<'_, BS>,
+        epoch: ChainEpoch,
+        partition_sectors: &PartitionSectorMap,
+        sector_size: SectorSize,
+        quant: QuantSpec,
+    ) -> Result<PowerPair, Box<dyn StdError>> {
+        let mut partitions = self.partitions_amt(store)?;
+
+        let mut power_lost = PowerPair::zero();
+        for (partition_idx, sector_numbers) in partition_sectors.iter() {
+            let mut partition = partitions
+                .get(partition_idx)
+                .map_err(|e| format!("failed to load partition {}: {:?}", partition_idx, e))?
+                .ok_or_else(
+                    || actor_error!(ErrNotFound; "failed to find partition {}", partition_idx),
+                )?;
+
+            let removed = partition
+                .terminate_sectors(store, sectors, epoch, sector_numbers, sector_size, quant)
+                .map_err(|e| {
+                    ActorError::downcast_wrap(
+                        e,
+                        format!("failed to terminate sectors in partition {}", partition_idx),
+                    )
+                })?;
+
+            partitions.set(partition_idx, partition).map_err(|e| {
+                format!(
+                    "failed to store updated partition {}: {:?}",
+                    partition_idx, e
+                )
+            })?;
+
+            if !removed.is_empty() {
+                // Record that partition now has pending early terminations.
+                self.early_terminations.set(partition_idx as usize);
+
+                // Record change to sectors and power
+                self.live_sectors -= removed.len() as u64;
+            } // note: we should _always_ have early terminations, unless the early termination bitfield is empty.
+
+            self.faulty_power -= &removed.faulty_power;
+
+            // Aggregate power lost from active sectors
+            power_lost += &removed.active_power;
+        }
+
+        // save partitions back
+        self.partitions = partitions
+            .flush()
+            .map_err(|e| format!("failed to persist partitions: {:?}", e))?;
```

Comment:
Hmm I'm noticing with these that there could be an out of gas error that would be ignored when translated like this, I think we will have to make a pass through and be able to downcast errors from amt and hamt flushes as well, since go uses the %w xerrors here.

I can open up an issue later, just noting one case for reference